### PR TITLE
[usm] `FileRegistry` improvements

### PR DESCRIPTION
**Pull request by p-lambert**
### What does this PR do?

* Simplify `FileRegistry` API
* Refactor `sharedlibraries.Watcher` tests to decouple them from implementation details;
* Add unit tests for `FileRegistry`
* Standardize the use of uint32 for PIDs everywhere (we were mixing int, int32, and uint32)

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/usm/utils/testutils.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
```
Comment (guyarb) [inline pkg/network/usm/utils/testutils.go:6]:
```suggestion
//go:build linux_bpf && test
```

File: pkg/network/usm/utils/testutils.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+package utils
+
+import "sync"
+
+type CallbackRecorder struct {
```
Comment (guyarb) [inline pkg/network/usm/utils/testutils.go:12]: please add documentation here

File: pkg/network/usm/sharedlibraries/testutil/file_opener.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/testutil/file_opener.go:5]: add a build flag `test`

File: pkg/network/usm/sharedlibraries/testutil/file_opener.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	return cmd
+}
+
+func build(t *testing.T) {
+	mux.Lock()
+	defer mux.Unlock()
+
+	if cachedArtifactPath != "" {
+		return
+	}
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/testutil/file_opener.go:44]: why?
document the function and the different assumptions you have

Comment (guyarb) [inline pkg/network/usm/sharedlibraries/testutil/file_opener.go:44]: you're implementing `sync.Once`, unless we have a good reason, I vote for moving into `sync.Once`

File: pkg/network/usm/sharedlibraries/testutil/file_opener.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+	"github.com/stretchr/testify/require"
+)
+
+// mutex protecting build process
+var mux sync.Mutex
+var cachedArtifactPath string
+
+func OpenFromAnotherProcess(t *testing.T, paths ...string) *exec.Cmd {
+	build(t)
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/testutil/file_opener.go:23]: the code cannot work in the CI, as we cannot build in the context of kitchen tests (missing go binary, etc.)
that's why in the original code we had a setup for the ci + building the sowatcher client in during `system-probe.kitchen-prepare`

File: pkg/network/usm/sharedlibraries/testutil/file_opener.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+	go func() {
+		<-sigs
+		done <- true
+	}()
+
+	fmt.Println("awaiting signal")
+	<-done
+	fmt.Println("exiting")
+}
+`
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/testutil/file_opener.go:104]: I don't see the value on having the "source" code like that. We cannot easily find issues in the find and cannot benefit from the use of an IDE to detect problems.

File: pkg/network/usm/sharedlibraries/testutil/file_opener.go
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/testutil/file_opener.go:1]: combining all of my comments - I think you need to restore to the original implementation of the sowatcher client.

File: pkg/network/usm/utils/file_registry_test.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+import (
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+
+	"github.com/DataDog/datadog-agent/pkg/network/protocols/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/network/usm/sharedlibraries/testutil"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry_test.go:19]:
```suggestion
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/DataDog/datadog-agent/pkg/network/protocols/telemetry"
	"github.com/DataDog/datadog-agent/pkg/network/usm/sharedlibraries/testutil"
```

File: pkg/network/usm/utils/file_registry.go [hunk 8]
```diff
@@ -8,30 +8,71 @@
...
+func NewFilePath(procRoot, namespacedPath string, pid uint32) (FilePath, error) {
+	// usw cwd of the process as root if the namespacedPath is relative
+	if namespacedPath[0] != '/' {
+		namespacedPath = "/cwd" + namespacedPath
+	}
+
+	path := fmt.Sprintf("%s/%d/root%s", procRoot, pid, namespacedPath)
+	pathID, err := NewPathIdentifier(path)
+	if err != nil {
+		fmt.Println("couldn't find file at", path)
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry.go:68]: is it a debugging leftover or should it become a trace log?

File: pkg/network/usm/utils/file_registry.go [hunk 9]
```diff
@@ -8,30 +8,71 @@
...
-type activationCB func(id PathIdentifier, root string, path string) error
-type deactivationCB func(id PathIdentifier) error
+// FilePath represents the location of a file from the *root* namespace view
+type FilePath struct {
+	HostPath string
+	ID       PathIdentifier
+}
+
+func NewFilePath(procRoot, namespacedPath string, pid uint32) (FilePath, error) {
+	// usw cwd of the process as root if the namespacedPath is relative
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry.go:60]:
```suggestion
	// use cwd of the process as root if the namespacedPath is relative
```

File: pkg/network/usm/utils/file_registry.go [hunk 10]
```diff
@@ -40,170 +81,160 @@ func NewFileRegistry() *FileRegistry {
...
+// Register inserts or updates a new file registration within to the `FileRegistry`;
+//
+// If no current registration exists for the given `PathIdentifier`, we execute
+// its *activation* callback. Otherwise, we increment the reference counter for
+// the existing registration if and only if `pid` is new;
+func (r *FileRegistry) Register(namespacedPath string, pid uint32, activationCB, deactivationCB callback) {
+	if activationCB == nil && deactivationCB == nil {
+		// if *both* callbacks are nil, there is no point in proceeding
+		return
+	}
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry.go:111]: I think it is a rare and unwanted scenario, so we should add a log about it.

File: pkg/network/usm/utils/file_registry.go [hunk 11]
```diff
@@ -40,170 +81,160 @@ func NewFileRegistry() *FileRegistry {
...
+
+	path, err := NewFilePath(r.procRoot, namespacedPath, pid)
 	if err != nil {
-		// short living process can hit here
-		// as we receive the openat() syscall info after receiving the EXIT netlink process
-		if log.ShouldLog(seelog.TraceLvl) {
-			log.Tracef("can't create path identifier %s", err)
-		}
 		return
 	}
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry.go:116]: why did you remove the log?

File: pkg/network/usm/utils/file_registry.go [hunk 12]
```diff
@@ -8,30 +8,71 @@
...
+// PID 50 opens /foobar => *activation* callback is executed; /foobar references=1
+// PID 60 opens /foobar => no callback is executed; /foobar references=2
+// PID 50 terminates => no callback is executed; /foobar references=1
+// PID 60 terminates => *deactivation* callback is executed; /foobar references=0
 type FileRegistry struct {
-	m     sync.RWMutex
-	byID  map[PathIdentifier]*registration
-	byPID map[uint32]pathIdentifierSet
+	m        sync.RWMutex
+	stopped  bool
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry.go:42]: what does it mean?

File: pkg/network/usm/utils/file_registry.go [hunk 13]
```diff
@@ -40,170 +81,160 @@ func NewFileRegistry() *FileRegistry {
...
-		if deactivate != nil {
-			if err := deactivate(pathID); err != nil {
-				log.Debugf("UnregisterCB library %s path %s : %s", pathID.String(), hostLibPath, err)
-			}
-		}
-		// save sentinel value, so we don't attempt to re-register shared
-		// libraries that are problematic for some reason
+	if err := activationCB(path); err != nil {
+		// we are calling `deactivationCB` here as some uprobes could be already attached
+		_ = deactivationCB(FilePath{ID: pathID})
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry.go:145]: there is not guarantee in your code that either `activationCB` or `deactivationCB` are not null (only that BOTH are not null)

File: pkg/network/usm/utils/file_registry.go (hunk 10 above)
Comment (guyarb) [reply pkg/network/usm/utils/file_registry.go:111]: furthermore, there is no point at registering if activeCB is nil.
if deactiveCB can be nil? TBH, I think not, but that's up to you.
thus you should change the code to be `if activationCB == nil {` or `if activationCB == nil || deactivationCB == nil {`

File: pkg/network/usm/utils/file_registry.go [hunk 14]
```diff
@@ -40,170 +81,160 @@ func NewFileRegistry() *FileRegistry {
...
+
+	_, found := r.byPID[pid]
 	if !found {
 		return
 	}
 
-	r.m.Lock()
-	defer r.m.Unlock()
-	paths, found := r.byPID[pidU32]
+	paths, found := r.byPID[pid]
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry.go:181]: now that you lock on the beginning of the file, this is a redundant duplication of line 176

File: pkg/network/usm/utils/file_registry.go [hunk 15]
```diff
@@ -40,170 +81,160 @@ func NewFileRegistry() *FileRegistry {
...
-	for pathID, reg := range r.byID {
-		reg.unregisterPath(pathID)
+// Clear removes all registrations calling their deactivation callbacks
+// This function should be called once during in termination.
+func (r *FileRegistry) Clear() {
+	r.m.Lock()
+	defer r.m.Unlock()
+	if r.stopped {
+		return
 	}
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry.go:219]: how that can happen?

File: pkg/network/usm/utils/file_registry_test.go (hunk 7 above)
Comment (guyarb) [inline pkg/network/usm/utils/file_registry_test.go:19]:
```suggestion
	"github.com/DataDog/datadog-agent/pkg/network/protocols/telemetry"
	"github.com/DataDog/datadog-agent/pkg/network/usm/sharedlibraries/testutil"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
```
```suggestion
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	
	"github.com/DataDog/datadog-agent/pkg/network/protocols/telemetry"
	"github.com/DataDog/datadog-agent/pkg/network/usm/sharedlibraries/testutil"
```

File: pkg/network/usm/utils/file_registry_test.go [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+
+	// Assert that the callback should executed only *once*
+	assert.Equal(t, 1, registerRecorder.CallsForPathID((pathID)))
+
+	// Assert that the two PIDs are being tracked
+	assert.Contains(t, r.GetRegisteredProcesses(), pid1)
+	assert.Contains(t, r.GetRegisteredProcesses(), pid2)
+
+	// Assert that the first call to `Unregister` (from PID 10) doesn't trigger
+	// the callback but removes the PID 10 from the list
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry_test.go:65]: PID 10?

File: pkg/network/usm/utils/file_registry_test.go [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+	path := filepath.Join(t.TempDir(), name)
+
+	f, err := os.Create(path)
+	require.NoError(t, err)
+	f.Close()
+
+	pathID, err := NewPathIdentifier(path)
+	require.NoError(t, err)
+
+	return path, pathID
```
Comment (guyarb) [inline pkg/network/usm/utils/file_registry_test.go:136]: you're leaking the files
please add
```go
t.Cleanup(func() {
    os.Remove(path)
})
```

File: pkg/network/usm/sharedlibraries/watcher_test.go [hunk 18]
```diff
@@ -94,42 +75,17 @@ func (s *SharedLibrarySuite) TestSharedLibraryDetection() {
 	launchProcessMonitor(t)
 
 	// create files
-	clientBin := buildSOWatcherClientBin(t)
-	command1 := exec.Command(clientBin, fooPath1)
-	require.NoError(t, command1.Start())
-	registerProcessTerminationUponCleanup(t, command1)
+	command1 := fileopener.OpenFromAnotherProcess(t, fooPath1)
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/watcher_test.go:78]: CI is broken now

File: pkg/network/usm/sharedlibraries/watcher_test.go [hunk 19]
```diff
@@ -94,42 +75,17 @@ func (s *SharedLibrarySuite) TestSharedLibraryDetection() {
...
-	require.GreaterOrEqual(t, tel["usm.so_watcher.hits"], tel["usm.so_watcher.matches"], "usm.so_watcher.hits")
-	telEqual(t, 0, "usm.so_watcher.already_registered")
-	telEqual(t, 0, "usm.so_watcher.blocked")
-	telEqual(t, 1, "usm.so_watcher.matches")
-	telEqual(t, 1, "usm.so_watcher.registered")
-	telEqual(t, 0, "usm.so_watcher.unregister_errors")
-	telEqual(t, 1, "usm.so_watcher.unregister_no_callback")
-	telEqual(t, 0, "usm.so_watcher.unregister_failed_cb")
-	telEqual(t, 0, "usm.so_watcher.unregister_pathid_not_found")
-	telEqual(t, 1, "usm.so_watcher.unregistered")
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/watcher_test.go:132]: why don't you check the telemetry?

File: pkg/network/usm/sharedlibraries/watcher_test.go [hunk 20]
```diff
@@ -186,26 +142,11 @@ func (s *SharedLibrarySuite) TestSharedLibraryDetectionWithPIDandRootNameSpace()
 	time.Sleep(10 * time.Millisecond)
 
 	// assert that soWatcher detected foo-libssl.so being opened and triggered the callback
-	require.Equal(t, libpath, pathDetected)
+	require.True(t, strings.HasSuffix(pathDetected, libpath))
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/watcher_test.go:145]: why did you switch from exact match to HasSuffix?

File: pkg/network/usm/sharedlibraries/watcher_test.go (hunk 19 above)
Comment (guyarb) [reply pkg/network/usm/sharedlibraries/watcher_test.go:132]: the comment is applied to the entire file

File: pkg/network/usm/sharedlibraries/watcher_test.go [hunk 21]
```diff
@@ -218,63 +159,36 @@ func (s *SharedLibrarySuite) TestSameInodeRegression() {
 	require.NoError(t, os.Link(fooPath1, fooPath2))
 	fooPathID2, err := utils.NewPathIdentifier(fooPath2)
 	require.NoError(t, err)
+	require.True(t, fooPathID1 == fooPathID2)
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/watcher_test.go:162]: you should use `Equal` instead, as the error message will give more details
```suggestion
	require.Equal(t, fooPathID1, fooPathID2)
```

File: pkg/network/usm/sharedlibraries/watcher_test.go [hunk 22]
```diff
@@ -303,76 +222,38 @@ func (s *SharedLibrarySuite) TestSoWatcherLeaks() {
...
 
 	require.NoError(t, command2.Process.Kill())
 	require.Eventuallyf(t, func() bool {
-		// Checking path1 still exists, and path2 not.
-		return checkPathIDDoesNotExist(watcher, fooPathID1) && checkPathIDDoesNotExist(watcher, fooPathID2)
+		// Checking that the unregisteredCB was executed now for pathID1
+		return unregisterRecorder.CallsForPathID(fooPathID1) == 1
 	}, time.Second*10, time.Second, "")
-
-	checkWatcherStateIsClean(t, watcher)
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/watcher_test.go:360]: why did you remove that?
The test is intended to detect leaks, but now we cannot detect them if they exist

**Review by guyarb — CHANGES_REQUESTED** [review]
CRed

There are areas that are more documented, and some that are not at all.
Please ensure the code is fully documented

File: pkg/network/usm/utils/file_registry.go (hunk 10 above)
Comment (p-lambert) [reply pkg/network/usm/utils/file_registry.go]: to be fair I don't think there is a real use case where either would be nil, so I'm requiring both to be non-nil now

File: pkg/network/usm/utils/file_registry.go (hunk 14 above)
Comment (p-lambert) [reply pkg/network/usm/utils/file_registry.go:181]: Good catch

File: pkg/network/usm/utils/file_registry.go (hunk 15 above)
Comment (p-lambert) [reply pkg/network/usm/utils/file_registry.go:219]: `Clear()` could be called multiple times

File: pkg/network/usm/utils/file_registry_test.go (hunk 17 above)
Comment (p-lambert) [reply pkg/network/usm/utils/file_registry_test.go:136]: I refactor the code to fix the CI problem, but I'm relying on `t.TempDir()` instead which has the same effect

File: pkg/network/usm/sharedlibraries/watcher_test.go (hunk 21 above)
Comment (p-lambert) [reply pkg/network/usm/sharedlibraries/watcher_test.go]: Good catch, left this by mistake

File: pkg/network/usm/sharedlibraries/testutil/file_opener.go (hunk 5 above)
Comment (p-lambert) [reply pkg/network/usm/sharedlibraries/testutil/file_opener.go]: I refactored the code and documented this, since it wasn't really clear in the code

File: pkg/network/usm/sharedlibraries/watcher_test.go (hunk 22 above)
Comment (p-lambert) [reply pkg/network/usm/sharedlibraries/watcher_test.go:360]: Re-added check by ensuring that are no PIDs registered

File: pkg/network/usm/sharedlibraries/watcher_test.go (hunk 19 above)
Comment (p-lambert) [reply pkg/network/usm/sharedlibraries/watcher_test.go:132]: Because telemetry no longer belongs to the `sharedlibraries.Watcher`

File: pkg/network/usm/sharedlibraries/watcher_test.go (hunk 20 above)
Comment (guyarb) [reply pkg/network/usm/sharedlibraries/watcher_test.go:149]: wondering if we should use regex match here to ensure specific pattern
```go
fmt.Sprintf("/proc/(+\d)/root/%", libpath)
```

File: pkg/network/usm/sharedlibraries/watcher_test.go [hunk 23]
```diff
@@ -403,117 +291,38 @@ func (s *SharedLibrarySuite) TestSoWatcherProcessAlreadyHoldingReferences() {
...
-		return checkPathIDDoesNotExist(watcher, fooPathID1) && checkPathIDDoesNotExist(watcher, fooPathID2)
+		// Assert that unregisterCB has also been called now for pathID1
+		return unregisterRecorder.CallsForPathID(fooPathID1) == 1 &&
+			unregisterRecorder.CallsForPathID(fooPathID2) == 1 &&
+			!hasPID(watcher, command1) &&
+			!hasPID(watcher, command2)
 	}, time.Second*10, time.Second, "")
-
-	checkWatcherStateIsClean(t, watcher)
-
```
Comment (guyarb) [inline pkg/network/usm/sharedlibraries/watcher_test.go:452]: should add

```go
	// Check there are no more processes registered
	assert.Len(t, watcher.registry.GetRegisteredProcesses(), 0)
```

**Comment by guyarb** [comment]
left 1 suggestion and 1 request

File: pkg/network/usm/sharedlibraries/watcher_test.go (hunk 20 above)
Comment (p-lambert) [reply pkg/network/usm/sharedlibraries/watcher_test.go:149]: I think it's fine to leave a simpler assertion here since I'm testing this more "in isolation" on the FileRegistry test suite:
https://github.com/DataDog/datadog-agent/pull/18355/files#diff-0d5db148a114a525cd8f679d11d07697be6bf6f1405733e97a9db66b2a39e1c8R108-R125

**Review by usamasaqib — APPROVED** [review]
Reviewed for ebpf-platform owned files.

**Review by guyarb — APPROVED** [review]

**Review by nplanel — APPROVED** [review]
